options/options.cpp

```diff
@@ -0,0 +1,199 @@
+ ** This file is part of the pono project.
+ ** Copyright (c) 2019, 2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file LICENSE in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief
+ **
+ **
+ **/
+
+#include "options/options.h"
+#include "optionparser.h"
+#include <vector>
+#include <iostream>
+#include <string>
+
+using namespace std;
+
+  /************************************* Option Handling setup
+ * *****************************************/
+// from optionparser-1.7 examples -- example_arg.cc
+enum optionIndex
+{
+  UNKNOWN_OPTION,
+  HELP,
+  ENGINE,
+  BOUND,
+  PROP,
+  VERBOSITY,
+  VCDNAME,
+  NOWITNESS
+};
+
+struct Arg : public option::Arg
+{
+  static void printError(const char * msg1,
+                         const option::Option & opt,
+                         const char * msg2)
+  {
+    fprintf(stderr, "%s", msg1);
+    fwrite(opt.name, opt.namelen, 1, stderr);
+    fprintf(stderr, "%s", msg2);
+  }
+
+  static option::ArgStatus Numeric(const option::Option & option, bool msg)
+  {
+    char * endptr = 0;
+    if (option.arg != 0 && strtol(option.arg, &endptr, 10)) {
+    };
+    if (endptr != option.arg && *endptr == 0) return option::ARG_OK;
+
+    if (msg) printError("Option '", option, "' requires a numeric argument\n");
+    return option::ARG_ILLEGAL;
+  }
+
+  static option::ArgStatus NonEmpty(const option::Option & option, bool msg)
+  {
+    if (option.arg != 0 && option.arg[0] != 0) return option::ARG_OK;
+
+    if (msg)
+      printError("Option '", option, "' requires a non-empty argument\n");
+    return option::ARG_ILLEGAL;
+  }
+};
+
+const option::Descriptor usage[] = {
+  { UNKNOWN_OPTION,
+    0,
+    "",
+    "",
+    Arg::None,
+    "USAGE: pono [options] <btor file>\n\n"
+    "Options:" },
+  { HELP, 0, "", "help", Arg::None, "  --help \tPrint usage and exit." },
+  { ENGINE,
+    0,
+    "e",
+    "engine",
+    Arg::NonEmpty,
+    "  --engine, -e <engine> \tSelect engine from [bmc, bmc-sp, ind, "
+    "interp]." },
+  { BOUND,
+    0,
+    "k",
+    "bound",
+    Arg::Numeric,
+    "  --bound, -k \tBound to check up until (default: 10)." },
+  { PROP,
+    0,
+    "p",
+    "prop",
+    Arg::Numeric,
+    "  --prop, -p \tProperty index to check (default: 0)." },
+  { VERBOSITY,
+    0,
+    "v",
+    "verbosity",
+    Arg::Numeric,
+    "  --verbosity, -v \tVerbosity for printing to standard out." },
+  { VCDNAME,
+    0,
+    "",
+    "vcd",
+    Arg::NonEmpty,
+    "  --vcd \tName of Value Change Dump (VCD) if witness exists." },
+  { NOWITNESS, 0, "", "no-witness", Arg::None, "  --no-witness \tDisable printing of witness." },
+  { 0, 0, 0, 0, 0, 0 }
+};
+/*********************************** end Option Handling setup
+ * ***************************************/
+
+namespace pono {
+
+bool PonoOptions::instance_created = false;
```

Comment:
This is global, right? Does this mean there's only one PonoOptions object allowed? We don't have to do this now, but it might actually be nice to allow multiple option configurations in memory. In IC3IA for example, options are a struct that's passed to the prover and it accesses the different options from there. Then, you can have programs that run provers with different options.